lib/trivial.nix

```diff
@@ -467,6 +471,41 @@ rec {
     then v
     else k: v;
 
+  /*
+    Make a function lazy in its arguments.
+
+    Normally when you call a function like `{ a, b }: { r = foo a b; }`,
+    the attribute names of the argument must be evaluated before the evaluator
+    can return a the (partially evaluated) function body (`{ r = ...; }`).
+
+    `lazyFunction` uses reflection on the passed in function and it only works
+    for functions defined with a "set pattern".
+
+    It makes a function like `{ a, b }: { r = foo a b; }` behave like
+    `args: { r = foo args.a args.b; }`.
+
+    It is not compatible with ellipsis patterns, `{ ... }:`, because only the
+    explicitly declared parameters are passed through. For example, the
+    following will not work: `lazyFunction(args@{ a, ... }: args.b)`;
+    even if the caller provides `b`.
+
+    It also removes the checks against unexpected arguments.
```

Comment:
```suggestion
    This function has some limitations:
    - It doesn't work for functions that don't do attribute destructuring like `arg: { r = foo arg.a arg.b; }` (`a`/`b` wouldn't be accessible)
    - It doesn't work for functions with ellipses (`...`) like `arg@{ a, ... }: { r = foo a arg.b; }` (`b` wouldn't be accessible)
    - It removes checks against unexpected arguments, so `lazyFunction ({ a }: a) { a = "a"; b = "b"; }` doesn't give an error.
```

Reply:
I'm just calling it attribute destructuring here instead of "set pattern", because nobody calls it that

Reply:
Somewhat jokingly I wrote https://github.com/NixOS/nixpkgs/pull/194992, which could check for some of these limitations in the code itself

Reply:
An author of the Nix manual uses this term.